### DISCUSSION TITLE
Preloading media in the background while participant is reading instructions?

### DISCUSSION THREAD

**Discussion by m-guseva** [jspsych/jsPsych#742] [Q&A] [answered]
Hello!
I have around 100 videos to preload for my experiment and I already figured out how to do it piecewise with this code (with the help of the google mailing list, thanks again for that!)

```
videos_vw1 = ['video1.mp4', 'video2.mp4']

var preload1 = {
        type: "call-function",
        async: true,
        func: function (done) {
            jsPsych.getDisplayElement().innerHTML = '<p> The experiment is loading</p>'
            jsPsych.pluginAPI.preloadVideo(videos_vw1, function () {
                done({ preload: "success" });
            })
        }
    };
```

I was wondering whether it would be possible to have this preloading procedure run **while** the participant is e.g. reading the instructions to avoid the long waiting times inbetween. Is that possible?
Note, I am not using the video keyboard response plug-in to display the video, but just a html-keyboard-response with the video specified in the stimulus.

Thank you!

**Comment by vijaymarupudi** [comment, marked as answer]
You can use this approach, which would require you to use the video-response- plugin though to load the preloaded video file. Otherwise, you'll need to implement a version of the html-keyboard-response plugin that does that.

There is a way to make this work using the [on_load](https://www.jspsych.org/overview/callbacks/#on_load) callback, but that would mess up your RT data.

```javascript
const makePreloadVideoPromise = function (videos) {
  return new Promise(function(resolve) {
    jsPsych.pluginAPI.preloadVideo(videos, resolve)
  })
}


// start loading
const videosPromise = makePreloadVideoPromise(["path/to/video1"])


const ensureVideosLoadedTrial = {
  type: "call-function",
  async: true,
  func: function(done) {
    videosPromise.then(function() { done({ "status": "success" }); })
  }
}


// after instructions
timeline.push(ensureVideosLoadedTrial)
```

**Reply by m-guseva** [reply]
Thank you Vijay, I'll mark this as an answer, however I realized that I have a more fundamental issue (I opened a new thread)

**Reply by zhangchi199133** [reply]
Hi @vijaymarupudi ! I wonder if I can chime in. I tried to replicate the background file preloading in timeline variables. Here is the code I have:

`        var preloadProcedure = {
              timeline_variables: trial_stimuli,
              timeline: [
                {
              type: "call-function",
              async: true,
              func: function (done) {
                const makePreloadAudioPromise = function (audios) {
                  return new Promise(function(resolve) {
                    jsPsych.pluginAPI.preloadAudioFiles(audios, resolve)
                  })
                };
                const audioPromise = makePreloadAudioPromise([jsPsych.timelineVariable('audio', true)]) 
                audioPromise.then(function() { done({ "status": "success" }); }
              )
              }
            }
          ]
        }

        timeline.push(preloadProcedure)`

But this seemed not working: the file preloading was still running in the foreground. I was wondering where my codes went wrong. Thank you so much!
Best,
Chi

**Reply by vijaymarupudi** [reply]
This code doesn't turn off regular preloading, you'll have to do that manually by passing an empty array to the preload_audio parameter of `jsPsych.init()`

Otherwise your approach should work, but it's rather inefficient, as it loads the files one after another instead of doing them all at once. If speed is not an issue, I would not change it though!

**Reply by zhangchi199133** [reply]
Hello Vijay! Thank you so much for the reply. I believe I have deactivated the automatic preloading by commented out the `preload_audio` argument in `jsPsych.init()`.

Speed is actually a huge (and the biggest) issue to me. The reason I use manual preloading instead of automatic preloading is that I had a large number of stimuli (image and audio) to preload, so I used the manual preloading to prevent the experiment from timeout before everything is loaded. But by using manual preloading it took more than five minutes to preload everything. So the option I can think of right now seems to be 1) to maunally load multiple files in parallel or 2) to run preloading in the background (i.e., the question of this thread). I actually have no idea how these two could be realized. Would you please give me some advice on that (or some other ideas to shorten the automatic or manual preloading time).

I am sorry for taking too much space in this thread. If it is necessary I will create a new thread.

Best,

Chi

**Reply by vijaymarupudi** [reply]
Yep, let's create a new thread and I can help you out!
